MIDAS/src/gnc/yessir.cpp

```diff
@@ -1,47 +1,344 @@
+        };
+        // TODO This mutex lock is almost certainly not necessary
+        //chMtxLock(&barometer.mutex);
+        sum += barometer.altitude;
+        //chMtxUnlock(&barometer.mutex);
+
+        //chMtxLock(&highG.mutex);
+        init_accel(0, 0) += accelerations.az;
+        init_accel(1, 0) += accelerations.ay;
+        init_accel(2, 0) += -accelerations.ax;
+        //chMtxUnlock(&highG.mutex);
+        THREAD_SLEEP(100);
+    }
+
+    init_accel(0, 0) /= 30;
+    init_accel(1, 0) /= 30;
+    init_accel(2, 0) /= 30;
+
+    //chMtxLock(&orientation.mutex);
+    euler_t euler = orientation.getEuler();
+    //chMtxUnlock(&orientation.mutex);
+    euler.yaw = -euler.yaw;
+    world_accel = BodyToGlobal(euler, init_accel);
+
+    // set x_k
+    x_k(0, 0) = sum / 30;
+    // x_k(0,0) = 1401;
+    x_k(3, 0) = 0;
+    x_k(6, 0) = 0;
+
+    // set F
+    for (int i = 0; i < 3; i++) {
+        F_mat(3 * i, 3 * i + 1) = s_dt;
+        F_mat(3 * i, 3 * i + 2) = (s_dt * s_dt) / 2;
+        F_mat(3 * i + 1, 3 * i + 2) = s_dt;
+
+        F_mat(3 * i, 3 * i) = 1;
+        F_mat(3 * i + 1, 3 * i + 1) = 1;
+        F_mat(3 * i + 2, 3 * i + 2) = 1;
+    }
+
+    Q(0, 0) = pow(s_dt, 5) / 20;
+    Q(0, 1) = pow(s_dt, 4) / 8;
+    Q(0, 2) = pow(s_dt, 3) / 6;
+    Q(1, 1) = pow(s_dt, 3) / 8;
+    Q(1, 2) = pow(s_dt, 2) / 2;
+    Q(2, 2) = s_dt;
+    Q(1, 0) = Q(0, 1);
+    Q(2, 0) = Q(0, 2);
+    Q(2, 1) = Q(1, 2);
 
-void Yessir::update() {}
+    Q(3, 3) = pow(s_dt, 5) / 20;
+    Q(3, 4) = pow(s_dt, 4) / 8;
+    Q(3, 5) = pow(s_dt, 3) / 6;
+    Q(4, 4) = pow(s_dt, 3) / 8;
+    Q(4, 5) = pow(s_dt, 2) / 2;
+    Q(5, 5) = s_dt;
+    Q(4, 3) = Q(3, 4);
+    Q(5, 3) = Q(3, 5);
+    Q(5, 4) = Q(4, 5);
 
-void Yessir::tick(float dt, float sd, Barometer &barometerData, Acceleration &imuData) {
-    // double integrate accelerometer data and store in state struct
-    state.acceleration = {
-        .ax = imuData.ax * 9.8f - 9.8f,
-        .ay = imuData.ay * 9.8f,
-        .az = imuData.az * 9.8f
-    };
+    Q(6, 6) = pow(s_dt, 5) / 20;
+    Q(6, 7) = pow(s_dt, 4) / 8;
+    Q(6, 8) = pow(s_dt, 3) / 6;
+    Q(7, 7) = pow(s_dt, 3) / 8;
+    Q(7, 8) = pow(s_dt, 2) / 2;
+    Q(8, 8) = s_dt;
+    Q(7, 6) = Q(6, 7);
+    Q(8, 6) = Q(6, 8);
+    Q(8, 7) = Q(7, 8);
 
-    state.velocity = {
-        .vx = state.velocity.vx + state.acceleration.ax * dt,
-        .vy = state.velocity.vy + state.acceleration.ay * dt,
-        .vz = state.velocity.vz + state.acceleration.az * dt
-    };
+    // set H
+    H(0, 0) = 1;
+    H(1, 2) = 1;
+    H(2, 5) = 1;
+    H(3, 8) = 1;
 
-    state.position = {
-        .px = state.position.px + state.velocity.vx * dt + 0.5f * state.acceleration.ax * dt * dt,
-        .py = state.position.py + state.velocity.vy * dt + 0.5f * state.acceleration.ay * dt * dt,
-        .pz = state.position.pz + state.velocity.vz * dt + 0.5f * state.acceleration.az * dt * dt
-    };
+    float spectral_density = 13.0;
```

Comment:
spectral density is already defined in yessir.h. Should it be defined in both places?